rcl/test/memory_tools_common.cpp

```diff
@@ -0,0 +1,159 @@
+  }
+  *unexpected_malloc_callback = callback;
+}
+
+void *
+custom_malloc(size_t size)
+{
+  if (!enabled.load()) {return malloc(size);}
+  auto foo = SCOPE_EXIT(enabled.store(true););
+  enabled.store(false);
+  if (!malloc_expected) {
+    if (unexpected_malloc_callback) {
+      (*unexpected_malloc_callback)();
+    }
+  }
+  void * memory = malloc(size);
+  uint64_t fw_size = size;
+  MALLOC_PRINTF(
+    "malloc  expected(%s): %p %" PRIu64 "\n",
+    malloc_expected ? "true " : "false", memory, fw_size);
+  return memory;
+}
+
+static __thread bool realloc_expected = true;
+static __thread UnexpectedCallbackType * unexpected_realloc_callback = nullptr;
+void set_on_unepexcted_realloc_callback(UnexpectedCallbackType callback)
+{
+  if (unexpected_realloc_callback) {
+    unexpected_realloc_callback->~UnexpectedCallbackType();
+    free(unexpected_realloc_callback);
+    unexpected_realloc_callback = nullptr;
+  }
+  if (!callback) {
+    return;
+  }
+  if (!unexpected_realloc_callback) {
+    unexpected_realloc_callback =
+      reinterpret_cast<UnexpectedCallbackType *>(malloc(sizeof(UnexpectedCallbackType)));
+    if (!unexpected_realloc_callback) {
+      throw std::bad_alloc();
+    }
+    new (unexpected_realloc_callback) UnexpectedCallbackType();
+  }
+  *unexpected_realloc_callback = callback;
+}
+
+void *
+custom_realloc(void * memory_in, size_t size)
+{
+  if (!enabled.load()) {return realloc(memory_in, size);}
+  auto foo = SCOPE_EXIT(enabled.store(true););
+  enabled.store(false);
+  if (!realloc_expected) {
+    if (unexpected_realloc_callback) {
+      (*unexpected_realloc_callback)();
+    }
+  }
+  void * memory = realloc(memory_in, size);
+  uint64_t fw_size = size;
+  MALLOC_PRINTF(
+    "realloc expected(%s): %p %p %" PRIu64 "\n",
+    malloc_expected ? "true " : "false", memory_in, memory, fw_size);
+  return memory;
+}
+
+static __thread bool free_expected = true;
+static __thread UnexpectedCallbackType * unexpected_free_callback = nullptr;
+void set_on_unepexcted_free_callback(UnexpectedCallbackType callback)
+{
+  if (unexpected_free_callback) {
+    unexpected_free_callback->~UnexpectedCallbackType();
+    free(unexpected_free_callback);
+    unexpected_free_callback = nullptr;
+  }
+  if (!callback) {
+    return;
+  }
+  if (!unexpected_free_callback) {
+    unexpected_free_callback =
+      reinterpret_cast<UnexpectedCallbackType *>(malloc(sizeof(UnexpectedCallbackType)));
+    if (!unexpected_free_callback) {
+      throw std::bad_alloc();
+    }
+    new (unexpected_free_callback) UnexpectedCallbackType();
+  }
+  *unexpected_free_callback = callback;
+}
+
+void
+custom_free(void * memory)
+{
+  if (!enabled.load()) {return free(memory);}
+  auto foo = SCOPE_EXIT(enabled.store(true););
+  enabled.store(false);
+  if (!free_expected) {
+    if (unexpected_free_callback) {
+      (*unexpected_free_callback)();
+    }
+  }
+  MALLOC_PRINTF("free    expected(%s): %p\n", malloc_expected ? "true " : "false", memory);
```

Comment:
That looks better. You could also get rid of the whitespace in the
parentheses and transfer it to after the close paren so that the address is
still aligned (maybe?) but then the "expected" tokens wouldn't line up.
 *shrug*

On Wed, Dec 16, 2015 at 4:05 PM, William Woodall <notifications@github.com>
wrote:

> In rcl/test/memory_tools_common.cpp
> <https://github.com/ros2/rcl/pull/5#discussion_r47852163>:
>
> > +  }
> > +  *unexpected_free_callback = callback;
> > +}
> > +
> > +void
> > +custom_free(void * memory)
> > +{
> > +  if (!enabled.load()) {return free(memory);}
> > +  auto foo = SCOPE_EXIT(enabled.store(true););
> > +  enabled.store(false);
> > +  if (!free_expected) {
> > +    if (unexpected_free_callback) {
> > +      (*unexpected_free_callback)();
> > +    }
> > +  }
> > +  MALLOC_PRINTF("free    expected(%s): %p\n", malloc_expected ? "true " : "false", memory);
>
> I ended up with 5372c52
> <https://github.com/ros2/rcl/commit/5372c520922f633aea2bf4884b239ae49d3da6ef>,
> which looks like this:
>
> 6: [==========] Running 1 test from 1 test case.
> 6: [----------] Global test environment set-up.
> 6: [----------] 1 test from TestMemoryTools
> 6: [ RUN      ] TestMemoryTools.test_allocation_checking_tools
> 6: test_memory_tools(32191,0x7fff7348f300) malloc: starting memory checking...
> 6: test_memory_tools(32191,0x7fff7348f300) malloc:  malloc (    expected) 0x7f82cb800000 1024
> 6: test_memory_tools(32191,0x7fff7348f300) malloc: realloc (    expected) 0x7f82cb800000 0x7f82cb800400 2048
> 6: test_memory_tools(32191,0x7fff7348f300) malloc:    free (    expected) 0x7f82cb800400
> 6: test_memory_tools(32191,0x7fff7348f300) malloc:  malloc (not expected) 0x7f82cb800000 1024
> 6: test_memory_tools(32191,0x7fff7348f300) malloc: realloc (    expected) 0x7f82cb800000 0x7f82cb800400 2048
> 6: test_memory_tools(32191,0x7fff7348f300) malloc:    free (    expected) 0x7f82cb800400
> 6: test_memory_tools(32191,0x7fff7348f300) malloc:  malloc (not expected) 0x7f82cb800000 1024
> 6: test_memory_tools(32191,0x7fff7348f300) malloc: realloc (    expected) 0x7f82cb800000 0x7f82cc000000 2048
> 6: test_memory_tools(32191,0x7fff7348f300) malloc:    free (    expected) 0x7f82cc000000
> 6: test_memory_tools(32191,0x7fff7348f300) malloc:  malloc (    expected) 0x7f82cc000800 1024
> 6: test_memory_tools(32191,0x7fff7348f300) malloc: realloc (    expected) 0x7f82cc000800 0x7f82cc000000 2048
> 6: test_memory_tools(32191,0x7fff7348f300) malloc:    free (    expected) 0x7f82cc000000
> 6: test_memory_tools(32191,0x7fff7348f300) malloc:  malloc (    expected) 0x7f82cc000800 1024
> 6: test_memory_tools(32191,0x7fff7348f300) malloc: realloc (    expected) 0x7f82cc000800 0x7f82cc000000 2048
> 6: test_memory_tools(32191,0x7fff7348f300) malloc:    free (    expected) 0x7f82cc000000
> 6: test_memory_tools(32191,0x7fff7348f300) malloc:  malloc (    expected) 0x7f82cc000800 1024
> 6: test_memory_tools(32191,0x7fff7348f300) malloc: realloc (    expected) 0x7f82cc000800 0x7f82cc000000 2048
> 6: test_memory_tools(32191,0x7fff7348f300) malloc:    free (    expected) 0x7f82cc000000
> 6: test_memory_tools(32191,0x7fff7348f300) malloc: stopping memory checking...
> 6: [       OK ] TestMemoryTools.test_allocation_checking_tools (1 ms)
> 6: [----------] 1 test from TestMemoryTools (1 ms total)
> 6:
> 6: [----------] Global test environment tear-down
> 6: [==========] 1 test from 1 test case ran. (1 ms total)
> 6: [  PASSED  ] 1 test.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/ros2/rcl/pull/5/files#r47852163>.
>